src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteOutputService.java

```diff
@@ -0,0 +1,475 @@
+      // remote output service to request the digest. The service may be
+      // able to return the digest from its bookkeeping, as opposed to
+      // reading the actual file contents.
+      BatchStatRequest.Builder builder = BatchStatRequest.newBuilder();
+      builder.setBuildId(currentBuildId.toString());
+      builder.setIncludeFileDigest(true);
+      builder.setFollowSymlinks(true);
+      PathFragment relativePath;
+      try {
+        relativePath = path.relativeTo(originalOutputPath);
+      } catch (IllegalArgumentException e) {
+        // Path is outside the output path. Send it to the regular file system.
+        return super.getFastDigest(path);
+      }
+      builder.addPaths(relativePath.toString());
+
+      BatchStatResponse responses = blockingStub.batchStat(builder.build());
+      StatResponse response = responses.getResponses(0);
+      if (!response.hasFileStatus()) {
+        // Output service reports that the file does not exist.
+        throw new FileNotFoundException();
+      }
+      FileStatus fileStatus = response.getFileStatus();
+      if (fileStatus.hasFile()) {
+        FileStatus.File regularFileStatus = fileStatus.getFile();
+        return DigestUtil.toBinaryDigest(regularFileStatus.getDigest());
+      }
+      if (fileStatus.hasExternal()) {
+        // Path is a symbolic link that points to a location outside of
+        // the output path.  The output service is unable to give a
+        // digest. We must compute it ourselves.
+        return super.getFastDigest(
+            originalOutputPath.getRelative(fileStatus.getExternal().getNextPath()));
+      }
+      throw new IOException(format("Remote output service did not return the status of file %s", path));
+    }
+  }
+
+  @Override
+  @Nullable
+  public FileSystem createActionFileSystem(
+      FileSystem sourceDelegate,
+      PathFragment execRootFragment,
+      String relativeOutputPath,
+      ImmutableList<Root> sourceRoots,
+      ActionInputMap inputArtifactData,
+      Iterable<Artifact> outputArtifacts,
+      boolean trackFailedRemoteReads) {
+    return new GrpcFileSystem(sourceDelegate, execRootFragment.getRelative(relativeOutputPath));
+  }
+
+  @Override
+  public void checkActionFileSystemForLostInputs(FileSystem actionFileSystem, Action action)
+      throws LostInputsActionExecutionException {
+    // There is no need to provide an implementation for this function,
+    // as StartBuild() on the server is supposed to call
+    // FindMissingBlobs() for all remotely tracked files. This ensures
+    // that they cannot disappear during the build.
+  }
+
+  public ListenableFuture<Void> downloadActionResult(ActionResult actionResult) {
+    // Request that all outputs of the action are created. Do make sure
+    // to remove the "bazel-out/" prefix from all paths, as that part of
+    // the path is not managed by the remote output service.
+    BatchCreateRequest.Builder builder = BatchCreateRequest.newBuilder();
```

Comment:
Internally, we don't batch these requests. I think that part of the reason is that the internal workflow looks something like this: (1) optimistically ask the local daemon if the file with given digest is already known, and if it is, insert it at this file path; (2) if not known, issue a save call to the remote execution service for each such output; (3) on that save call's (asynchronous) completion do an insert. This avoids the expense of a remote RPC and save to backends if the file is already known locally, which is a fairly common case. I don't know if that complexity is needed for the initial implementation of the Bazel protocol.

The incentive to start with the more complex solution is for the proto definition here to essentially be the open-sourced internal version, allowing for more code sharing, but there may well be a lot of path dependence in how we got to where we are. Perhaps we can try to make something that shares as much as possible without burdening Bazel with premature complexity.

Reply:
For the FUSE daemon that I created, BatchCreate() is always a cheap call. It never loads any objects from the CAS. It just creates placeholder files in the file system that are backed by CAS objects. Only when opening the file and reading one or more bytes of data, it will fault in the file from the remote execution service and store it in the local system's cache. This means that it's 'valid' to create files that use a Digest that doesn't correspond to an actual CAS object. There is no need to call to the remote execution service to save the file. REv2 also doesn't provide any facilities for that, right?

That said, if someone wants to write an implementation that explicitly calls into the remote execution service to save files, then there is nothing that prevents that, right? It's just that it's the responsibility of the FUSE daemon. It can do that as part of its BatchCreate() implementation.

Reply:
I'm not advocating for the control flow that Blaze uses, but to clarify, it's Blaze's responsibility to issue that remote RPC to save the outputs: it is already in communication with the remote service, and at present, I'm not sure that the local insertion call has enough data for the FUSE daemon to issue the proper request. I know nothing about REv2, so if it has no facility to explicitly save outputs, then I agree that Bazel is unlikely to be pushed in the direction of individual inserts versus this batch approach.

Reply:
I am also in the camp of not very knowledgeable about REv2. Does REv2 facilitate both permanent storage of outputs and execution? If so, how do we ensure that the file remains available in case I do `bazel build //something; bazel shutdown` and browse the FUSE with my favorite tool?

The optimistic insert Janak mentioned works as an optimization where if the output store already knows a digest, there is no need to store the same thing again -- we may as well reuse that. I guess if the FUSE/NFS merely registers the metadata on how to get the file from RE when user/local action requests it, then such optimization will not buy us anything here.